metricbeat/helper/prometheus/textparse.go

```diff
@@ -0,0 +1,694 @@
+		}
+		quant.Quantile = &f
+		quant.Value = &s
+	}
+
+	_, k := summariesByName[name]
+	if !k {
+		summariesByName[name] = make(map[string]*OpenMetric)
+	}
+	metric, ok := summariesByName[name][lbls]
+	if !ok {
+		metric = &OpenMetric{}
+		metric.Name = &name
+		metric.Summary = summary
+		metric.Summary.Quantile = quantile
+		summariesByName[name][lbls] = metric
+	}
+	if metric.Summary.SampleSum == nil && summary.SampleSum != nil {
+		metric.Summary.SampleSum = summary.SampleSum
+	} else if metric.Summary.SampleCount == nil && summary.SampleCount != nil {
+		metric.Summary.SampleCount = summary.SampleCount
+	} else if quant.Quantile != nil {
+		metric.Summary.Quantile = append(metric.Summary.Quantile, quant)
+	}
+
+	return name, metric
+}
+
+func histogramMetricName(name string, s float64, qv string, lbls string, t *int64, isGaugeHistogram bool, e *exemplar.Exemplar, histogramsByName map[string]map[string]*OpenMetric) (string, *OpenMetric) {
+	var histogram = &Histogram{}
+	var bucket = []*Bucket{}
+	var bkt = &Bucket{}
+
+	switch {
+	case isCount(name):
+		u := uint64(s)
+		histogram.SampleCount = &u
+		name = name[:len(name)-6]
+	case isSum(name):
+		histogram.SampleSum = &s
+		name = name[:len(name)-4]
+	case isGaugeHistogram && isGCount(name):
+		u := uint64(s)
+		histogram.SampleCount = &u
+		name = name[:len(name)-7]
+	case isGaugeHistogram && isGSum(name):
+		histogram.SampleSum = &s
+		name = name[:len(name)-5]
+	default:
+		if isBucket(name) {
+			name = name[:len(name)-7]
+		}
+		f, err := strconv.ParseFloat(qv, 64)
+		if err != nil {
+			f = math.MaxUint64
+		}
+		cnt := uint64(s)
+		bkt.UpperBound = &f
+		bkt.CumulativeCount = &cnt
+
+		if e != nil {
+			if !e.HasTs {
+				e.Ts = *t
+			}
+			bkt.Exemplar = e
+		}
+	}
+
+	_, k := histogramsByName[name]
+	if !k {
+		histogramsByName[name] = make(map[string]*OpenMetric)
+	}
+	metric, ok := histogramsByName[name][lbls]
+	if !ok {
+		metric = &OpenMetric{}
+		metric.Name = &name
+		metric.Histogram = histogram
+		metric.Histogram.Bucket = bucket
+		histogramsByName[name][lbls] = metric
+	}
+	if metric.Histogram.SampleSum == nil && histogram.SampleSum != nil {
+		metric.Histogram.SampleSum = histogram.SampleSum
+	} else if metric.Histogram.SampleCount == nil && histogram.SampleCount != nil {
+		metric.Histogram.SampleCount = histogram.SampleCount
+	} else if bkt.UpperBound != nil {
+		metric.Histogram.Bucket = append(metric.Histogram.Bucket, bkt)
+	}
+
+	return name, metric
+}
+
+func ParseMetricFamilies(b []byte, contentType string, ts time.Time) ([]*MetricFamily, error) {
+	var (
+		parser               = textparse.New(b, contentType)
+		defTime              = timestamp.FromTime(ts)
+		metricFamiliesByName = map[string]*MetricFamily{}
+		summariesByName      = map[string]map[string]*OpenMetric{}
+		histogramsByName     = map[string]map[string]*OpenMetric{}
+		fam                  *MetricFamily
+		mt                   = textparse.MetricTypeUnknown
+	)
+	var err error
+
+loop:
+	for {
```

Comment:
Nit. This for loop does too many things, consider splitting it in smaller functions.

Reply:
Good catch! I'm leaning towards filing a separate issue for such improvements so as to ship the update of the library first and ensure that we don't mix changes that could be more error prone. wdyt?

Reply:
Sure thing, no need to do it now.